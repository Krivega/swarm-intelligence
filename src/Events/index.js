/**
 * Handle Error not supported
 *
 * @param {string} eventName - eventName
 *
 * @return {void}
 */
const errorNotSupported = eventName => new Error(`Event ${eventName} not supported`);

/**
 * Events
 *
 * @param {string} eventName - hack for fix validate jsdoc
 *
 * @returns {@class} Events - hack for fix validate jsdoc
 */
class Events {
  _active = true;

  _eventHandlers = {};

  _triggers = {};

  /**
   * @constructor
   *
   * @param {Object}   events        - The events
   * @param {Object}   options       - The options
   * @param {function} options.debug - The debug function
   */
  constructor(events, { debug } = {}) {
    this._events = events;
    this._debug = debug;

    this._initEventHandlers(this._events);
  }

  /**
   * on
   *
   * @param {string}   eventName - eventName
   * @param {function} handler   - handler
   *
   * @returns {void}
   */
  on(eventName, handler) {
    const handlers = this._eventHandlers[eventName];

    if (handlers) {
      handlers.push(handler);
    } else {
      throw errorNotSupported(eventName);
    }
  }

  /**
   * once
   *
   * @param {string}   eventName - eventName
   * @param {function} handler   - handler
   *
   * @returns {void}
   */
  once(eventName, handler) {
    /**
     * onceHandler
     *
     * @param {Object} args - args
     *
     * @returns {void}
     */
    const onceHandler = (...args) => {
      handler(...args);
      this.off(eventName, onceHandler);
    };

    this.on(eventName, onceHandler);
  }

  /**
   * off
   *
   * @param {string}   eventName - eventName
   * @param {function} handler   - handler
   *
   * @returns {void}
   */
  off(eventName, handler) {
    const handlers = this._eventHandlers[eventName];

    this._eventHandlers[eventName] = handlers.filter(item => item !== handler);
  }

  /**
   * trigger
   *
   * @param {string} eventName - eventName
   * @param {Object} data      - data
   *
   * @returns {void}
   */
  trigger(eventName, data) {
    const trigger = this._triggers[eventName];

    if (trigger) {
      trigger(data);
    } else {
      throw errorNotSupported(eventName);
    }
  }

  /**
   * triggers
   *
   * @returns {Object} triggers
   */
  get triggers() {
    return this._triggers;
  }

  /**
   * eachTriggers
   *
   * @param {function} handler - handler
   *
   * @returns {void}
   */
  eachTriggers(handler) {
    Object.entries(this._triggers).forEach(([eventName, trigger]) => {
      handler(trigger, eventName);
    });
  }

  /**
   * Removes event handlers.
   *
   * @returns {void}
   */
  removeEventHandlers() {
    this._initEventHandlers(this._events);
  }

  /**
   * _initEventHandlers
   *
   * @param {Object} eventsNames - eventsNames
   *
   * @returns {void}
   */
  _initEventHandlers(eventsNames) {
    eventsNames.forEach(eventName => {
      this._eventHandlers[eventName] = [];
      this._triggers[eventName] = this._resolveHandleEvent(eventName);
    });
  }

  /**
   * activate
   *
   * @returns {undefined}
   */
  activate() {
    this._active = true;
  }

  /**
   * deactivate
   *
   * @returns {undefined}
   */
  deactivate() {
    this._active = false;
  }

  _resolveHandleEvent = eventName => (...args) => {
    if (!this._active) {
      return undefined;
    }

    const eventHandlers = this._eventHandlers[eventName];

    eventHandlers.forEach(eventHandler => {
      try {
        eventHandler(...args);
      } catch (error) {
        if (this._debug) {
          this._debug(error);
        } else {
          throw error;
        }
      }
    });

    return undefined;
  };
}

export default Events;
