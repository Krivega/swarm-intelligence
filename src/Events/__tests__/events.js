import Events from '../index';

const initEventNames = ['event1', 'event2'];
const eventNames = ['event3', 'event4'];
const [, eventName] = initEventNames;

let debug;

let events;

let mockFn;

let arg;

describe('events', () => {
  beforeEach(() => {
    debug = jest.fn();
    events = new Events(initEventNames, { debug });
    mockFn = jest.fn();
    arg = {};
  });

  it('on', () => {
    events.on(eventName, mockFn);
    expect(events._eventHandlers[eventName][0]).toBe(mockFn);
    expect(events).toMatchSnapshot();
  });

  it('once', () => {
    const mockTriggerOnceFn = jest.fn();
    const mockTriggerOnceFn2 = jest.fn();

    events.once(eventName, mockTriggerOnceFn);
    events.once(eventName, mockTriggerOnceFn2);
    events.on(eventName, mockFn);

    expect(events._eventHandlers[eventName].length).toBe(3);

    events.trigger(eventName, arg);

    expect(events._eventHandlers[eventName].length).toBe(1);
    expect(mockTriggerOnceFn).toHaveBeenCalledTimes(1);
    expect(mockTriggerOnceFn2).toHaveBeenCalledTimes(1);

    events.trigger(eventName, arg);

    expect(mockTriggerOnceFn).toHaveBeenCalledTimes(1);
    expect(mockTriggerOnceFn2).toHaveBeenCalledTimes(1);
    expect(mockFn).toHaveBeenCalledTimes(2);
    expect(events._eventHandlers[eventName].length).toBe(1);
    expect(events._eventHandlers[eventName][0]).toBe(mockFn);
    expect(events).toMatchSnapshot();
  });

  it('handle errors', () => {
    const mockTriggerOnceFn = jest.fn(() => {
      throw new Error('Error');
    });
    const mockTriggerOnceFn2 = jest.fn();

    events.once(eventName, mockTriggerOnceFn);
    events.once(eventName, mockTriggerOnceFn2);
    events.on(eventName, mockFn);

    expect(events._eventHandlers[eventName].length).toBe(3);

    events.trigger(eventName, arg);

    expect(events._eventHandlers[eventName].length).toBe(2);
    expect(debug).toHaveBeenCalledTimes(1);
    expect(mockTriggerOnceFn).toHaveBeenCalledTimes(1);
    expect(mockTriggerOnceFn2).toHaveBeenCalledTimes(1);

    events.trigger(eventName, arg);

    expect(mockTriggerOnceFn).toHaveBeenCalledTimes(2);
    expect(mockTriggerOnceFn2).toHaveBeenCalledTimes(1);
    expect(mockFn).toHaveBeenCalledTimes(2);
    expect(events._eventHandlers[eventName].length).toBe(2);
    expect(events._eventHandlers[eventName][1]).toBe(mockFn);
    expect(events).toMatchSnapshot();
  });

  it('off', () => {
    events.on(eventName, mockFn);
    events.off(eventName, mockFn);
    expect(events._eventHandlers[eventName].length).toBe(0);
    expect(events).toMatchSnapshot();
  });

  it('removeEventHandlers', () => {
    events.on(eventName, mockFn);
    events.removeEventHandlers();
    expect(events._eventHandlers[eventName].length).toBe(0);
    expect(events).toMatchSnapshot();
  });

  it('trigger', () => {
    events.on(eventName, mockFn);
    events.trigger(eventName, arg);
    events.trigger(eventName, arg);

    expect(mockFn).toHaveBeenCalledTimes(2);
    expect(mockFn).toBeCalledWith(arg);
    expect(events).toMatchSnapshot();
  });

  it('get triggers', () => {
    expect(events.triggers).toHaveProperty(initEventNames[0]);
    expect(events._triggers).toHaveProperty(initEventNames[1]);
    expect(events).toMatchSnapshot();
  });

  it('eachTriggers', () => {
    events.eachTriggers(mockFn);

    expect(mockFn).toHaveBeenCalledTimes(initEventNames.length);
    expect(mockFn.mock.calls[0][0]).toEqual(expect.any(Function));
    expect(mockFn.mock.calls[0][1]).toBe(initEventNames[0]);
    expect(mockFn.mock.calls[1][0]).toEqual(expect.any(Function));
    expect(mockFn.mock.calls[1][1]).toBe(initEventNames[1]);
    expect(events).toMatchSnapshot();
  });

  it('_initEventHandlers', () => {
    events._initEventHandlers(eventNames);
    expect(events._triggers).toHaveProperty(eventNames[0]);
    expect(events._triggers).toHaveProperty(eventNames[1]);
    expect(events).toMatchSnapshot();
  });

  it('_resolveHandleEvent', () => {
    events.on(eventName, mockFn);
    events._resolveHandleEvent(eventName)(arg);

    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(mockFn).toBeCalledWith(arg);
    expect(events).toMatchSnapshot();
  });

  it('error with debug', () => {
    events.on(eventName, () => {
      throw new Error('test');
    });
    events.trigger(eventName, arg);

    expect(debug).toHaveBeenCalledTimes(1);
  });

  it('error without debug', () => {
    events = new Events(initEventNames);

    const error = new Error('test');

    expect(() => {
      events.on(eventName, () => {
        throw error;
      });
      events.trigger(eventName, arg);
    }).toThrow(error);
  });

  it('on not supported error', () => {
    const eventNameNotSupported = 'eventName not supported';

    expect(() => {
      events.on(eventNameNotSupported, mockFn);
    }).toThrow(new Error(`Event ${eventNameNotSupported} not supported`));
  });

  it('trigger not supported error', () => {
    const eventNameNotSupported = 'eventName not supported';

    expect(() => {
      events.trigger(eventNameNotSupported, mockFn);
    }).toThrow(new Error(`Event ${eventNameNotSupported} not supported`));
  });

  it('activate/deactivate', () => {
    events.on(eventName, mockFn);
    events.trigger(eventName, arg);
    events.trigger(eventName, arg);

    expect(mockFn).toHaveBeenCalledTimes(2);
    expect(mockFn).toBeCalledWith(arg);

    events.deactivate();
    events.trigger(eventName, arg);
    events.trigger(eventName, arg);

    expect(mockFn).toHaveBeenCalledTimes(2);

    events.activate();
    events.trigger(eventName, arg);
    events.trigger(eventName, arg);

    expect(mockFn).toHaveBeenCalledTimes(4);

    expect(events).toMatchSnapshot();
  });
});
