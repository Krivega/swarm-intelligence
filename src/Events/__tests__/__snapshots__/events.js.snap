// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`events _initEventHandlers 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [],
    "event3": Array [],
    "event4": Array [],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
    "event3": [Function],
    "event4": [Function],
  },
}
`;

exports[`events _resolveHandleEvent 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    ],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;

exports[`events activate/deactivate 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    ],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;

exports[`events eachTriggers 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;

exports[`events get triggers 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;

exports[`events handle errors 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction] {
    "calls": Array [
      Array [
        [Error: Error],
      ],
      Array [
        [Error: Error],
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [
      [Function],
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    ],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;

exports[`events off 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;

exports[`events on 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [
      [MockFunction],
    ],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;

exports[`events once 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    ],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;

exports[`events removeEventHandlers 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;

exports[`events trigger 1`] = `
Events {
  "_active": true,
  "_debug": [MockFunction],
  "_eventHandlers": Object {
    "event1": Array [],
    "event2": Array [
      [MockFunction] {
        "calls": Array [
          Array [
            Object {},
          ],
          Array [
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    ],
  },
  "_events": Array [
    "event1",
    "event2",
  ],
  "_resolveHandleEvent": [Function],
  "_triggers": Object {
    "event1": [Function],
    "event2": [Function],
  },
}
`;
